### PR TITLE
Extra docs

### PR DESCRIPTION
**Pull request by derekdreery**
I always try to understand a codebase by writing docs, here they are.

### CODE REVIEW

File: crates/backend/src/ast.rs [hunk 1]
```diff
@@ -36,7 +36,9 @@ pub struct Export {
     pub constructor: Option<String>,
     /// The rust function
     pub function: Function,
+    ///
```
Comment (fitzgen) [inline crates/backend/src/ast.rs]: We should either not have a doc comment at all, or the comment should have some content.

File: crates/backend/src/ast.rs [hunk 2]
```diff
@@ -36,7 +36,9 @@ pub struct Export {
...
     pub comments: Vec<String>,
+    /// The name of the rust object the function belongs to TODO is this correct?
```
Comment (fitzgen) [inline crates/backend/src/ast.rs]: This is the name of the rust function/method, not the object. Check out `impl ToTokens for ast::Export` to see how it is used.

File: crates/macro/src/parser.rs [hunk 3]
```diff
@@ -44,6 +48,7 @@ impl BindgenAttrs {
             .next()
     }
 
+    /// Get the first module attribute
```
Comment (fitzgen) [inline crates/macro/src/parser.rs]: s/module/version/

File: crates/backend/src/ast.rs [hunk 4]
```diff
@@ -255,6 +275,7 @@ impl Export {
     }
 }
 
+
```
Comment (fitzgen) [inline crates/backend/src/ast.rs]: Nitpick: remove this newly introduced empty line.

File: crates/backend/src/ast.rs [hunk 5]
```diff
@@ -230,6 +249,7 @@ impl Export {
         Ident::new(&generated_name, Span::call_site())
     }
 
+    /// ???
```
Comment (fitzgen) [inline crates/backend/src/ast.rs]: This is the name of the shim function that gets exported and takes the raw ABI form of its arguments and converts them back into their normal, "high level" form before calling the actual function.

File: crates/macro/src/parser.rs (hunk 3 above)
Comment (derekdreery) [reply crates/macro/src/parser.rs]: oops

**Comment by derekdreery** [comment]
Ready for review again.

File: crates/webidl/src/first_pass.rs [hunk 6]
```diff
@@ -1,3 +1,12 @@
+//! Because some WebIDL constructs are defined in multiple places
+//! (keyword `partial` is used to add to an existing construct),
+//! We need to first walk the webidl to collect all non-partial
+//! constructs so that we have containers in which to put the
+//! partial ones.
+//!
+//! Only `interface`s, `dictionary`s, `enum`s and `mixin`s can
+//! be partial.
+
```
Comment (ohanar) [inline crates/webidl/src/first_pass.rs]: A couple of more things that have been on my mind and should go in the first pass in the future (likely not exhaustive):
- `typedef`s should not be exposed on the rust side, but should be processed by the binding's generator (i.e. this crate). To do that, `typedef` declarations have to be recorded before building the AST, so in the first pass.
- should we want to implement up-casting conversion by more than one level, we need to be able to walk the inheritance tree up to the root while building the AST, doing so would require recording that in the first pass

More generally, the fist pass is used to record information that needs to be globally accessible while building the backend AST.

**Pull request merged**